### ISSUE TITLE
[Bug] Tensor.exp() 無數值溢位保護，大輸入直接產生 inf，後續反向傳播出現 NaN

### ISSUE THREAD

**Issue by s9213712** [s9213712/minicnn#47]
## 問題描述

`src/minicnn/nn/tensor.py` 的 `exp()`：

```python
def exp(self) -> 'Tensor':
    out_data = np.exp(self.data)   # ← 無任何保護
    # ...
    def _backward() -> None:
        if out.grad is not None:
            self._add_grad(out.grad * out_data)  # ← inf * grad = inf 或 nan
```

當輸入值大於約 709（float64 的 exp 上限），`np.exp()` 回傳 `inf`。而 `inf * gradient` 在某些情況下產生 `nan`，使梯度傳播失效（梯度消失或爆炸到 NaN）。

## 觸發場景

1. **sigmoid 中的 exp**：雖然 sigmoid 使用 `1 / (1 + np.exp(-data))`，對正輸入不會溢位，但透過 `Tensor.exp()` 直接呼叫時可能溢位
2. **softmax 分母**：若 logits 很大（但 `log_softmax` 已有正確的 max-subtraction 保護），`cross_entropy` 路徑可能繞過保護
3. **自訂層**：使用者直接呼叫 `x.exp()` 時，無任何保護

## 建議修復

加入輸入值的 clamp：

```python
def exp(self) -> 'Tensor':
    clipped = np.clip(self.data, -500, 500)   # float32 安全範圍約 ±88，float64 約 ±709
    out_data = np.exp(clipped)
    out = Tensor(out_data, requires_grad=_requires_grad(self))
    out._prev = {self}
    out._op = 'exp'

    def _backward() -> None:
        if out.grad is not None:
            self._add_grad(out.grad * out_data)

    out._backward = _backward
    return out
```

注意：clamp 會使梯度不準確（clipped 區域梯度應為 0），建議同時在 backward 中：
```python
# 在 clipped 區域（輸入 > 500 或 < -500），exp 的梯度理論上仍是 exp(x)，但 clamp 破壞了此關係
# 嚴格做法：記錄 mask，clipped 區域梯度置 0
mask = (self.data >= -500) & (self.data <= 500)
self._add_grad(out.grad * out_data * mask)
```
